src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatDefaultPayload.cs

```diff
@@ -42,6 +100,84 @@ public static void PopulateDefaultPayload(IEnumerable<string> disabledFields, IH
+                                    .ConfigureAwait(false);
+
+                    var enabledImdsFields = enabledFields.Intersect(allFields);
+                    foreach (string field in enabledImdsFields)
+                    {
+                        heartbeatManager.AddHeartbeatProperty(
+                            propertyName: field,
+                            propertyValue: await GetAzureInstanceMetadaValue(baseImdsUrl, field, imdsApiVersion, imdsTextFormat)
+                                .ConfigureAwait(false),
+                            isHealthy: true,
+                            allowDefaultFields: true);
+                    }
+                }
+                catch (AggregateException)
+                {
+                    CoreEventSource.Log.CannotObtainAzureInstanceMetadata();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets all the available fields from the imds link asynchronously, and returns them as an IEnumerable.
+        /// </summary>
+        /// <param name="baseUrl">Url of the Azure Instance Metadata service</param>
+        /// <param name="apiVersion">rest Api version to append to the constructed Uri</param>
+        /// <param name="textFormatArg">query-string argument to request text data be returned</param>
+        /// <returns>an array of field names available, or null</returns>
+        private static async Task<IEnumerable<string>> GetAzureInstanceMetadataFields(string baseUrl, string apiVersion, string textFormatArg)
+        {
+            string allComputeFields = string.Empty;
+            string allComputeFieldsUrl = $"{baseUrl}?{textFormatArg}&{apiVersion}";
+
+            using (var httpClient = new HttpClient())
+            {
+                httpClient.DefaultRequestHeaders.Add("Metadata", "True");
+                allComputeFields = await httpClient.GetStringAsync(allComputeFieldsUrl).ConfigureAwait(false);
+            }
+
+            return allComputeFields.Split(Environment.NewLine.ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
+        }
+
+        /// <summary>
+        /// Gets the value of a specific field from the imds link asynchronously, and returns it.
+        /// </summary>
+        /// <param name="baseUrl">Url of the Azure Instance Metadata service</param>
+        /// <param name="fieldName">Specific imds field to retrieve a value for</param>
+        /// <param name="apiVersion">rest Api version to append to the constructed Uri</param>
+        /// <param name="textFormatArg">query-string argument to request text data be returned</param>
+        /// <returns>an array of field names available, or null</returns>
+        private static async Task<string> GetAzureInstanceMetadaValue(string baseUrl, string fieldName, string apiVersion, string textFormatArg)
+        { 
+            string computFieldUrl = $"{baseUrl}/{fieldName}?{textFormatArg}&{apiVersion}";
```

Comment:
I thought you can get a single JSON that has both - names and values. Do you really need to run an extra http request for every name?

Reply:
I am not convinced it's worthwhile to write a JSON parser here, and I doubt we want to take a dependency on something like Newtonsoft either. I chose this route as it removed the necessity for both options altogether. While I believe the JSON would likely be trivial to parse, I cannot guarantee the absence of edge cases that would cause us (and our customers, especially in regions where special characters are common in text strings) grief.